scripts/watcher.js

```diff
@@ -1,69 +1,121 @@
-#!/usr/bin/env node
 import watcher from '@parcel/watcher';
-import path from 'path';
-import fs from 'fs';
-import os from 'os';
+import fs from 'node:fs';
+import path from 'node:path';
+import { mapCatalogToAstro } from './map-catalog-to-astro.js';
+import { rimrafSync } from 'rimraf';
 
-// Where the users project is located
-const projectDirectory = process.env.PROJECT_DIR || process.cwd();
-// Where the catalog code is located.
-const catalogDirectory = process.env.CATALOG_DIR;
+/**
+ * @typedef {Object} Event
+ * @property {string} path
+ * @property {"create"|"update"|"delete"} type
+ *
+ * @typedef {(err: Error | null, events: Event[]) => unknown} SubscribeCallback
+ */
 
-const contentPath = path.join(catalogDirectory, 'src', 'content');
-
-const watchList = ['domains', 'commands', 'events', 'services', 'teams', 'users', 'pages', 'components', 'flows'];
-// const absoluteWatchList = watchList.map((item) => path.join(projectDirectory, item));
+/**
+ *
+ * @param {string} projectDirectory
+ * @param {string} catalogDirectory
+ * @param {SubscribeCallback|undefined} callback
+ */
+export async function watch(projectDirectory, catalogDirectory, callback = undefined) {
+  const subscription = await watcher.subscribe(
+    projectDirectory,
+    compose(
+      /**
+       * @param {Error|null} err
+       * @param {Event[]} events
+       * @returns {unknown}
+       */
+      (err, events) => {
+        if (err) {
+          return;
+        }
 
-// confirm folders exist before watching them
-const verifiedWatchList = watchList.filter((item) => fs.existsSync(path.join(projectDirectory, item)));
+        for (let event of events) {
+          const { path: filePath, type } = event;
 
-const extensionReplacer = (collection, file) => {
-  if (collection === 'teams' || collection == 'users') return file;
-  return file.replace('.md', '.mdx');
-};
+          const astroPaths = mapCatalogToAstro({
+            filePath,
+            astroDir: catalogDirectory,
+            projectDir: projectDirectory,
+          });
 
-for (let item of [...verifiedWatchList]) {
-  // Listen to the users directory for any changes.
-  watcher.subscribe(path.join(projectDirectory, item), (err, events) => {
-    if (err) {
-      return;
+          for (const astroPath of astroPaths) {
+            switch (type) {
+              case 'create':
+              case 'update':
+                if (fs.statSync(filePath).isDirectory()) fs.mkdirSync(astroPath, { recursive: true });
+                else retryEPERM(fs.cpSync)(filePath, astroPath);
+                break;
+              case 'delete':
+                retryEPERM(rimrafSync)(astroPath);
+                break;
+            }
+          }
+        }
+      },
+      callback
+    ),
+    {
+      ignore: [catalogDirectory],
```

Comment:
When running the code locally `npm run start:catalog` no changes are being watched. I think this is ignoring the whole directory?

Reply:
Cool I changed this, to still include the catalog directory but not to ignore files inside the project directory in the case the project DIR is inside the catalog dir, in this case for our local dev environment.